ESY-1986 Add section for MOTO payments in other features (#671)
SwedbankPay/developer.swedbankpay.com: _includes/moto.md

{% if include.api_resource == "creditcard" %}
    {% assign api_resource_field_name = "payment" %}
    {% assign api_redirect_rel = "authorization" %}
    {% assign api_redirect_link = "creditcard/payments/authorize" %}
{% else %}
    {% assign api_resource_field_name = "paymentorder" %}

>     {% assign api_redirect_rel = "paymentorder" %}

Reviewer: Is the name of the `rel` really `redirect-paymentorder`?

Reviewer 2: yes, I got this confirmed with @sigridbra

Author: Yes 😅

    "state": "Ready",
    "operation": "Purchase",
    "intent": "Authorization",
    "currency": "NOK",
    "amount": 1500,
    "remainingCaptureAmount": 1500,
    "remainingCancellationAmount": 1500,
    "remainingReversalAmount": 0,
    "description": "Test Moto",
    "initiatingSystemUserAgent": "PostmanRuntime/3.0.1",
    "userAgent": "Mozilla/5.0...",
    "language": "nb-NO",
    "prices": { "id": "/psp/{{ api_resource }}/payments/{{ page.payment_id }}/prices" },
    "transactions": { "id": "/psp/{{ api_resource }}/payments/{{ page.payment_id }}/transactions" },
    "authorizations": { "id": "/psp/{{ api_resource }}/payments/{{ page.payment_id }}/authorizations" },
    "captures": { "id": "/psp/{{ api_resource }}/payments/{{ page.payment_id }}/captures" },
    "reversals": { "id": "/psp/{{ api_resource }}/payments/{{ page.payment_id }}/reversals" },
    "cancellations": { "id": "/psp/{{ api_resource }}/payments/{{ page.payment_id }}/cancellations" },
    "urls" : { "id": "/psp/{{ api_resource }}/payments/{{ page.payment_id }}/urls" },
    "payeeInfo" : { "id": "/psp/{{ api_resource }}/payments/{{ page.payment_id }}/payeeInfo" }
  },
  "operations": [

>               {

Reviewer: ```suggestion
        {
```

>             "rel": "redirect-{{api_redirect_rel}}",

Reviewer: ```suggestion
            "rel": "redirect-{{ api_redirect_rel }}",
```

>             "href": "{{ page.front_end_url }}/{{api_redirect_link}}/{{ page.payment_token }}",

Reviewer: ```suggestion
            "href": "{{ page.front_end_url }}/{{ api_redirect_link }}/{{ page.payment_token }}",
```

            "method": "GET",
            "contentType": "text/html"
        },
  ]
}

> ```

Reviewer: @marinafrydenb, can you please add some text after this code example explaining that the `operation` with `rel` equal to `redirect-{{ api_redirect_rel }}` should be found and that the merchant employee needs to visit that page to fill out the payer's card details?

Reviewer 3: You should be able to find the `operation `with `rel `equal to `redirect-{{ api_redirect_rel }}` in the response. This is the link to the payment page. The merchant employee needs to visit this page and enter the card details provided by the payer.

Do we need more?

Reviewer 3: Sorry for hijacking your task, @marinafrydenb :|

Reviewer 2: no worries :)

Reviewer 2: loving all the help i can get :)

Reviewer 2: included this in the text :)

{% if include.api_resource == "creditcard" %}
    {% assign api_resource_field_name = "payment" %}
    {% assign api_redirect_rel = "authorization" %}
    {% assign api_redirect_link = "creditcard/payments/authorize" %}
{% else %}
    {% assign api_resource_field_name = "paymentorder" %}
    {% assign api_redirect_rel = "paymentorder" %}
    {% assign api_redirect_link = "paymentmenu" %}
{% endif %}

## MOTO

{% include alert-agreement-required.md %}

MOTO (Mail Order / Telephone Order) is a purchase where you as a merchant enter 
the payer's card details in order to make a payment. The card details are given
over telephone or in writing, and then filled out to the payment interface by the
sales representative. 

> Common use-cases are travel or hotel bookings, where the 
> payer calls the sales representative to make a booking. This feature is only 
> supported in the `Purchase` operation. See the abbreviated example below on how to
> implement MOTO by setting the `generateMotoPayment` to `true`. 

Reviewer: ```suggestion

Common use-cases are travel or hotel bookings, where the payer calls the sales
representative to make a booking. This feature is only supported in the
`Purchase` operation. See the abbreviated example below on how to implement MOTO
by setting the `generateMotoPayment` to `true`. 
```

 

{:.code-view-header}
**Request**

```http
POST /psp/{{ include.api_resource }}/payments HTTP/1.1
Host: {{ page.api_host }}
Authorization: Bearer <AccessToken>
Content-Type: application/json

{
    "{{api_resource_field_name}}": {
        "operation": "Purchase",
        "intent": "Authorization",
        "generateMotoPayment": true,
        "currency": "NOK",
        "amount": 1500,
        "vatAmount": 0,
        "description": "Test Moto",
        "userAgent": "Mozilla/5.0...",
        "language": "nb-NO",
        "urls": {
            "callbackUrl": "https://example.com/payment-callback"
        },
        "payeeInfo": {
            "payeeId": "{{ page.merchant_id }}",
            "payeeReference": "CD1234",
            "payeeName": "Merchant1",
            "productCategory": "A123",
            "orderReference": "or-12456",
            "subsite": "MySubsite"
        }
    }
}
```

To fill out the card information, find the operation with 
`rel` equal to `redirect-{{ api_redirect_rel }}` and redirect to the

> page where the merchant employee fills out the payer’s card details. 

Reviewer: ```suggestion
To authorization the payment, find the operation with `rel` equal to
`redirect-{{ api_redirect_rel }}` in the response and redirect the merchant
employee to the provided `href` to fill out the payer’s card details. You will
find an abbreviated example of the response provided below.
```

> ## MOTO

Reviewer 4: Codacy found an issue: [Command failed: git rev-parse --show-cdup
/bin/sh: git: not found
](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

{% include alert-agreement-required.md %}

Mail-Order Telephone-Order
Notes to ourself: 

>   * only supported by Purchase (not Verify), 

Reviewer 4: Codacy found an issue: [[list-item-bullet-indent] Incorrect indentation before bullet: remove 2 spaces](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

>   * Only valid with 3DS 2.x

Reviewer 4: Codacy found an issue: [[list-item-bullet-indent] Incorrect indentation before bullet: remove 2 spaces](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

>   * skips SCA

Reviewer 4: Codacy found an issue: [[list-item-bullet-indent] Incorrect indentation before bullet: remove 2 spaces](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

>   * Should be Type = MOTO and acquirerTransactionType = MOTO on the response for GET payment under payment/transactions/{authorize}/activities/{authorize}

Reviewer 4: Codacy found an issue: [[list-item-bullet-indent] Incorrect indentation before bullet: remove 2 spaces](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

> ## MOTO

Reviewer 4: Codacy found an issue: [Command failed: git rev-parse --show-cdup
/bin/sh: git: not found
](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

{% include alert-agreement-required.md %}

Mail-Order Telephone-Order
Notes to ourself: 

> - only supported by Purchase (not Verify), 

Reviewer 4: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

> - Only valid with 3DS 2.x

Reviewer 4: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

> - skips SCA

Reviewer 4: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

> - Should be Type = MOTO and acquirerTransactionType = MOTO on the response for GET payment under payment/transactions/{authorize}/activities/{authorize}

Reviewer 4: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

> {:.code-header}

Reviewer 4: Codacy found an issue: [[no-consecutive-blank-lines] Remove 1 line before node](https://app.codacy.com/gh/SwedbankPay/developer.swedbankpay.com/pullRequest?prid=6175237)

{% if include.api_resource == "creditcard" %}
    {% assign api_resource_field_name = "payment" %}
{% else %}
    {% assign api_resource_field_name = "paymentorder" %}
{% endif %}

## MOTO

{% include alert-agreement-required.md %}

MOTO (Mail Order / Telephone Order) is a purchase where you as a merchant enter 

> the customer card details in order to make a payment. The card details are given

Reviewer 3: ```suggestion
the payer's card details in order to make a payment. The card details are given
```
Just a bit of nitpicking to avoid confusion :) In the developer portal, we use the term _payer_ for _consumer_ / _end user_ / payer_. (Internally, _customer_ is the term used for _merchant_, and consumer is the term used for _payer/consumer_.)

over telephone or in writing, and then entered into the payment interface by the
sales representative. Common use-cases are travel or hotel bookings, where the 

> customer calls the sales representative to make a booking. This feature is only 

Reviewer 3: ```suggestion
payer calls the sales representative to make a booking. This feature is only 
```
See above :)

Reviewer 3: Great job! 💯

supported in the `Purchase` operation. See the abbreviated example below on how to

> implement MOTO by setting the `generatedMotoPayment` to `true`.

Reviewer 3: ```suggestion
implement MOTO by setting the `generateMotoPayment` to `true`.
```
A "d" had slipped in :)

 

{:.code-view-header}
**Request**

```http
POST /psp/{{ include.api_resource }}/payments HTTP/1.1
Host: {{ page.api_host }}
Authorization: Bearer <AccessToken>
Content-Type: application/json

{
    "{{api_resource_field_name}}": {
        "operation": "Purchase",
        "intent": "Authorization",
        "generateMotoPayment": true,
        "currency": "NOK",
        "amount": 1500,
        "vatAmount": 0,
        "description": "Test Moto",
        "userAgent": "Mozilla/5.0...",
        "language": "nb-NO",
        "urls": {
            "callbackUrl": "https://example.com/payment-callback"
        },
        "payeeInfo": {
            "payeeId": "{{ page.merchant_id }}",
            "payeeReference": "CD1234",
            "payeeName": "Merchant1",
            "productCategory": "A123",
            "orderReference": "or-12456",
            "subsite": "MySubsite"
        }
    }
}
```

{:.code-view-header}
**Response**

```http

> HTTP/1.1 200 OK

Reviewer: I don't quite understand what should happen after this response has been received. Where should the card details given from the payer over telephone be entered by the merchant employee? Is there a redirect URL to look for in the response that should be presented to the merchant employee or what?

Author: Yes! The merchant should use the redirect-operation.

Reviewer: Then we need to provide that operation in the example and explain what to do with it. :)

Reviewer 2: @sigridbra can you add en explanation in this PR ?

Author: This would be just as a normal purchase, the only difference being that the details are filled in by the sales representative.

> {% assign api_resource=include.api_resource %}

Reviewer: ```suggestion
```

{% if include.api_resource == "creditcard" %}
    {% assign api_resource_field_name = "payment" %}
{% else %}
    {% assign api_resource_field_name = "paymentorder" %}
{% endif %}

> ## MOTO (Mail Order / Telephone Order)

Reviewer: ```suggestion
## MOTO
```

{% include alert-agreement-required.md %}

MOTO (Mail Order / Telephone Order) is a purchase where you as a merchant enter 
the customer card details in order to make a payment. The card details are given
over telephone or in writing, and then entered into the payment interface by the
sales representative. Common use-cases are travel or hotel bookings, where the 
customer calls the sales representative to make a booking. This feature is only 
supported in the `Purchase` operation. See the abbreviated example below on how to
implement MOTO by setting the `generatedMotoPayment` to `true`.
 

{:.code-view-header}
**Request**

```http

> POST /psp/{{ api_resource }}/payments HTTP/1.1

Reviewer: ```suggestion
POST /psp/{{ include.api_resource }}/payments HTTP/1.1
```

{% assign api_resource=include.api_resource %}
{% if include.api_resource == "creditcard" %}
    {% assign api_resource_field_name = "payment" %}
{% else %}
    {% assign api_resource_field_name = "paymentorder" %}
{% endif %}

## MOTO (Mail Order / Telephone Order)

{% include alert-agreement-required.md %}

MOTO (Mail Order / Telephone Order) is a purchase where you as a merchant enter 
the customer card details in order to make a payment. The card details are given
over telephone or in writing, and then entered into the payment interface by the
sales representative. Common use-cases are travel or hotel bookings, where the 
customer calls the sales representative to make a booking. This feature is only 
supported in the `Purchase` operation and is only valid if you have 3-D Secure 2

> authentication of the cardholder. See the abbreviated example below on how to

Author: This might be misleading. Moto is only available if you have 3-D Secure 2 authentication activated in the contract. The purchase will never prompt a Strong Customer Authentication on the Cardholder for MOTO (as the cardholder is not present).

{% if include.api_resource == "creditcard" %}
    {% assign api_resource_field_name = "payment" %}

>     {% assign api_redirect_rel = "authorization" %}

Author: Is this valid in jekyll?

Reviewer: Yes. 👍
